### PR TITLE
Create all streams before copying in exec/attach

### LINKED ISSUE #16373: Exec/attach should create all streams before copying data
We create the stdin stream and immediately start copying local stdin to the container. We do the same for stdout and stderr (create stream, start copying immediately). In the kubelet, we wait to receive all the streams before we start the exec and copy between the streams and the container.

The spdy connection frame handling code has 5 frame workers, each with a queue of up to 50 frames. In the event that we try to push more than 50 frames to a single queue before all streams are ready, the 51st attempt to push a frame blocks until there is space available in the queue. But there won't be any space until the server starts reading from the streams, which won't happen until all the streams have been received, so we're stuck at this point.

The fix is to create all the streams first, and only start copying after all the streams have been created and the server is ready to copy between the streams and the container.

Originally reported here: https://github.com/openshift/origin/issues/5377

### PR DESCRIPTION
**Pull request by ncdc**
Create error, stdin, stdout, stderr streams first, and only start
copying once all the streams have been created. This fixes an issue
where the client immediately starts sending data for stdin before all
the other streams have been created. This ends up blocking the spdy
connection frame handler and causes the entire exec/attach session to
time out.

Fixes #16373

### CODE REVIEW

**Comment by ncdc** [comment]
@lavalamp @smarterclayton @liggitt @csrwng 

Note, I would like to try to write an e2e that fails w/o this commit, and works with it. I'll work on that later today.

File: pkg/client/unversioned/remotecommand/v1.go [hunk 1]
```diff
@@ -55,12 +55,50 @@ func (e *streamProtocolV1) stream(conn httpstream.Connection) error {
...
+	// set up all the streams first
 	headers := http.Header{}
 	headers.Set(api.StreamType, api.StreamTypeError)
-	errorStream, err := conn.CreateStream(headers)
+	errorStream, err = conn.CreateStream(headers)
 	if err != nil {
 		return err
 	}
+	defer errorStream.Reset()
+
```
Comment (liggitt) [inline pkg/client/unversioned/remotecommand/v1.go]: add a comment about why we set up all the streams first for posterity

**Comment by liggitt** [comment]
LGTM, a test would be good if possible

**Comment by csrwng** [comment]
LGTM, solved the issue with attaching stdin in Windows

**Comment by lavalamp** [comment]
Looks good, I'd also like a test. Are you wanting to get this into 1.1? I think I'd rather not kill ourselves to get this in and just plan on CPing into 1.1.1, unless you see some horrible consequences from that.

**Comment by ncdc** [comment]
I'm fine with waiting until 1.1.1.

**Comment by ncdc** [comment]
And I do plan to write a test, as soon as I fix another higher priority bug I'm working on.

**Comment by ncdc** [comment]
Unrelated failure:

```
Summarizing 1 Failure:

[Fail] Services [It] should be able to change the type and nodeport settings of a service 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:517
```

@k8s-bot test this

**Comment by lavalamp** [comment]
@ncdc Shall we wait for the test or do you want to send that separately?

**Comment by ncdc** [comment]
I'm working on a P0 issue now for OpenShift. I plan to attempt to do the test after the issue is resolved. I'm ok waiting to merge if you are.

**Comment by lavalamp** [comment]
Sure, waiting means we won't forget :)

**Comment by lavalamp** [comment]
Any updates?

**Comment by ncdc** [comment]
@lavalamp I could have sworn I added a comment last week, but I guess not...

I attempted to write a test that streamed the kubectl binary from the local system to a container. I could never get it to fail. Unless we can come up with other test ideas, I think this might just need to be merged...

File: pkg/client/unversioned/remotecommand/v1.go (hunk 1 above)
Comment (lavalamp) [inline pkg/client/unversioned/remotecommand/v1.go]: +1

Add this and I'll LGTM

Comment (ncdc) [inline pkg/client/unversioned/remotecommand/v1.go]: Updated, PTAL

**Comment by lavalamp** [comment]
LGTM

**Pull request merged**